lib/cn/cn_tree.c

```diff
@@ -911,19 +920,20 @@ cn_node_comp_token_put(struct cn_tree_node *tn)
 static void
 cn_comp_release(struct cn_compaction_work *w)
 {
+    bool join = (w->cw_action == CN_ACTION_JOIN);
     assert(w->cw_node);
 
-    if (w->cw_err) {
+    if (w->cw_err || join) {
         struct kvset_list_entry *le;
-        uint kx;
+        uint kx, cnt = join ? cn_ns_kvsets(&w->cw_node->tn_ns) : w->cw_kvset_cnt;
 
         /* unmark input kvsets */
-        le = w->cw_mark;
-        for (kx = 0; kx < w->cw_kvset_cnt; kx++) {
+        le = join ? list_last_entry(&w->cw_node->tn_kvset_list, typeof(*le), le_link) : w->cw_mark;
+        for (kx = 0; kx < cnt; kx++) {
             assert(le);
             assert(kvset_get_workid(le->le_kvset) != 0);
             kvset_set_workid(le->le_kvset, 0);
-            le = list_prev_entry(le, le_link);
+            le = list_prev_entry_or_null(le, le_link, &w->cw_node->tn_kvset_list);
```

Comment:
Define kx on a separate line to make cnt easier to read..

Reply:
Re: cw_mark,
cn_comp_release() is called after a join operation.
If the join was successful, then w->cw_node would also contain the kvsets moved from w->cw_join (assuming an non-empty w->cw_join node). So, cw_mark may not point to the last kvset any more.
That reminds me, I need to reset the workid for the kvsets in w->cw_join in the case of a failed join operation.